cmd/apprepository-controller/main.go

```diff
@@ -30,29 +35,94 @@ import (
+	RepoSyncCommand          string
+	Namespace                string
+	DBURL                    string
+	DBUser                   string
+	DBName                   string
+	DBSecretName             string
+	DBSecretKey              string
+	UserAgentComment         string
+	Crontab                  string
+	ReposPerNamespace        bool
+
+	// Args are the positional (non-flag) command-line arguments.
+	Args []string
+}
+
+// parseFlags parses the command-line arguments provided to the program.
+// Typically os.Args[0] is provided as 'programName' and os.args[1:] as 'args'.
+func parseFlags(progname string, args []string) (config *Config, output string, err error) {
+	flagSet := flag.NewFlagSet(progname, flag.ContinueOnError)
+	var buf bytes.Buffer
+	flagSet.SetOutput(&buf)
+
+	var conf Config
+
+	flagSet.StringVar(&conf.Kubeconfig, "kubeconfig", "", "Path to a kubeconfig. Only required if out-of-cluster.")
+	flagSet.StringVar(&conf.MasterURL, "master", "", "The address of the Kubernetes API server. Overrides any value in kubeconfig. Only required if out-of-cluster.")
+	flagSet.StringVar(&conf.RepoSyncImage, "repo-sync-image", "quay.io/helmpack/chart-repo:latest", "container repo/image to use in CronJobs")
+	flagSet.Var(&conf.RepoSyncImagePullSecrets, "repo-sync-image-pullsecrets", "optional reference to secrets in the same namespace to use for pulling the image used by this pod")
+	flagSet.StringVar(&conf.RepoSyncCommand, "repo-sync-cmd", "/chart-repo", "command used to sync/delete repos for repo-sync-image")
+	flagSet.StringVar(&conf.Namespace, "namespace", "kubeapps", "Namespace to discover AppRepository resources")
+	flagSet.BoolVar(&conf.ReposPerNamespace, "repos-per-namespace", true, "Defaults to watch for repos in all namespaces. Switch to false to watch only the configured namespace.")
+	flagSet.StringVar(&conf.DBURL, "database-url", "localhost", "Database URL")
+	flagSet.StringVar(&conf.DBUser, "database-user", "root", "Database user")
+	flagSet.StringVar(&conf.DBName, "database-name", "charts", "Database name")
+	flagSet.StringVar(&conf.DBSecretName, "database-secret-name", "kubeapps-db", "Kubernetes secret name for database credentials")
+	flagSet.StringVar(&conf.DBSecretKey, "database-secret-key", "postgresql-root-password", "Kubernetes secret key used for database credentials")
+	flagSet.StringVar(&conf.UserAgentComment, "user-agent-comment", "", "UserAgent comment used during outbound requests")
+	flagSet.StringVar(&conf.Crontab, "crontab", "*/10 * * * *", "CronTab to specify schedule")
+
+	err = flagSet.Parse(args)
+	if err != nil {
+		return nil, buf.String(), err
+	}
+	conf.Args = flagSet.Args()
+	return &conf, buf.String(), nil
+}
 
 func main() {
-	flag.Parse()
+	conf, output, err := parseFlags(os.Args[0], os.Args[1:])
+
+	if err != nil {
+		log.Fatal("got error:", err)
+		log.Fatal("output:\n", output)
+		os.Exit(1)
+	}
+
+	log.WithFields(log.Fields{
+		"kubeconfig":                  conf.Kubeconfig,
```

Comment:
make sure this doesn't logs secrets (like the serviceaccount token in the kubeconfig). I am not sure if this is just the path to the kubeconfig file or the actual content.

Reply:
No, AFAIK it is just the path to the kubeconfig file (`/home/agamez/.kube/config ` in my case). In fact, it is calling BuildConfigFromFlags method:
` BuildConfigFromFlags(masterUrl, kubeconfigPath string) (*restclient.Config, error)` which only needs the path, not the file itself.

The log trace I'm getting is:

`DEBUG[0000] apprepository-controller configured with these args:  crontab="*/10 * * * *" database-name=charts database-secret-key=postgresql-root-password database-secret-name=kubeapps-db database-url=localhost database-user=root kubeconfig=/home/agamez/.kube/config master= namespace=default repo-sync-cmd=/chart-repo repo-sync-image="demo.goharbor.io/docker/bitnami/kubeapps-asset-syncer:2.0.0-scratch-r2" repo-sync-image-pullsecrets="[harbor]" repos-per-namespace=true user-agent-comment=`

It does trace the secret names and the DB info, but just because this is the required input for the command to work.
However, it is just a debug log, I don't see it's a big deal, but we can directly remove it or change it to a _trace_ level.

Reply:
:+1: thanks for double checking!